From 3fe52a2eb9487b3baa16268442ba3bfe06a5f8fc Mon Sep 17 00:00:00 2001
From: Tomasz Nowicki <tn@semihalf.com>
Date: Mon, 28 Mar 2022 18:14:51 +0000
Subject: [PATCH 1/1] MANATEE: HACK: kvm: Fix IOAPIC default routing setup

In-kernel KVM irqchip is created on guest demand and IOAPIC is one of
the emulated IRQ controller component which is initialized at very
early stage. By default, KVM is using 1:1 GSI routing for all available
IOAPIC pin range which is 0-24 now.

Since the IOAPIC pin limit was increased to 120 at maximum,
any IRQ within 25-120 pin is silently dripped until routing table
is filled up by VMM. Fill in missing routing entries to cover full
0-120 pin range.

BUG=b:225020663
TEST=boot manatee and check if IRQs within 24-120 are working properly

Signed-off-by: Tomasz Nowicki <tn@semihalf.com>
Change-Id: I1bbaf8743bf0823a1df5c134a7c432a4f2e24c44
---
 arch/x86/kvm/irq_comm.c | 48 +++++++++++++++++++++++++++++++++++++++++
 1 file changed, 48 insertions(+)

diff --git a/arch/x86/kvm/irq_comm.c b/arch/x86/kvm/irq_comm.c
index 4aa1c2e00e2a..39cd03678fe3 100644
--- a/arch/x86/kvm/irq_comm.c
+++ b/arch/x86/kvm/irq_comm.c
@@ -373,6 +373,54 @@ static const struct kvm_irq_routing_entry default_routing[] = {
 	ROUTING_ENTRY1(18), ROUTING_ENTRY1(19),
 	ROUTING_ENTRY1(20), ROUTING_ENTRY1(21),
 	ROUTING_ENTRY1(22), ROUTING_ENTRY1(23),
+	ROUTING_ENTRY1(24), ROUTING_ENTRY1(25),
+	ROUTING_ENTRY1(26), ROUTING_ENTRY1(27),
+	ROUTING_ENTRY1(28), ROUTING_ENTRY1(29),
+	ROUTING_ENTRY1(30), ROUTING_ENTRY1(31),
+	ROUTING_ENTRY1(32), ROUTING_ENTRY1(33),
+	ROUTING_ENTRY1(34), ROUTING_ENTRY1(35),
+	ROUTING_ENTRY1(36), ROUTING_ENTRY1(37),
+	ROUTING_ENTRY1(38), ROUTING_ENTRY1(39),
+	ROUTING_ENTRY1(40), ROUTING_ENTRY1(41),
+	ROUTING_ENTRY1(42), ROUTING_ENTRY1(43),
+	ROUTING_ENTRY1(44), ROUTING_ENTRY1(45),
+	ROUTING_ENTRY1(46), ROUTING_ENTRY1(47),
+	ROUTING_ENTRY1(48), ROUTING_ENTRY1(49),
+	ROUTING_ENTRY1(50), ROUTING_ENTRY1(51),
+	ROUTING_ENTRY1(52), ROUTING_ENTRY1(53),
+	ROUTING_ENTRY1(54), ROUTING_ENTRY1(55),
+	ROUTING_ENTRY1(56), ROUTING_ENTRY1(57),
+	ROUTING_ENTRY1(58), ROUTING_ENTRY1(59),
+	ROUTING_ENTRY1(60), ROUTING_ENTRY1(61),
+	ROUTING_ENTRY1(62), ROUTING_ENTRY1(63),
+	ROUTING_ENTRY1(64), ROUTING_ENTRY1(65),
+	ROUTING_ENTRY1(66), ROUTING_ENTRY1(67),
+	ROUTING_ENTRY1(68), ROUTING_ENTRY1(69),
+	ROUTING_ENTRY1(70), ROUTING_ENTRY1(71),
+	ROUTING_ENTRY1(72), ROUTING_ENTRY1(73),
+	ROUTING_ENTRY1(74), ROUTING_ENTRY1(75),
+	ROUTING_ENTRY1(76), ROUTING_ENTRY1(77),
+	ROUTING_ENTRY1(78), ROUTING_ENTRY1(79),
+	ROUTING_ENTRY1(80), ROUTING_ENTRY1(81),
+	ROUTING_ENTRY1(82), ROUTING_ENTRY1(83),
+	ROUTING_ENTRY1(84), ROUTING_ENTRY1(85),
+	ROUTING_ENTRY1(86), ROUTING_ENTRY1(87),
+	ROUTING_ENTRY1(88), ROUTING_ENTRY1(89),
+	ROUTING_ENTRY1(90), ROUTING_ENTRY1(91),
+	ROUTING_ENTRY1(92), ROUTING_ENTRY1(93),
+	ROUTING_ENTRY1(94), ROUTING_ENTRY1(95),
+	ROUTING_ENTRY1(96), ROUTING_ENTRY1(97),
+	ROUTING_ENTRY1(98), ROUTING_ENTRY1(99),
+	ROUTING_ENTRY1(100), ROUTING_ENTRY1(101),
+	ROUTING_ENTRY1(102), ROUTING_ENTRY1(103),
+	ROUTING_ENTRY1(104), ROUTING_ENTRY1(105),
+	ROUTING_ENTRY1(106), ROUTING_ENTRY1(107),
+	ROUTING_ENTRY1(108), ROUTING_ENTRY1(109),
+	ROUTING_ENTRY1(110), ROUTING_ENTRY1(111),
+	ROUTING_ENTRY1(112), ROUTING_ENTRY1(113),
+	ROUTING_ENTRY1(114), ROUTING_ENTRY1(115),
+	ROUTING_ENTRY1(116), ROUTING_ENTRY1(117),
+	ROUTING_ENTRY1(118), ROUTING_ENTRY1(119),
 };
 
 int kvm_setup_default_irq_routing(struct kvm *kvm)
-- 
2.35.1.1094.g7c7d902a7c-goog

